### PR TITLE
Decompile DRA func_8010FDF8

### PR DESCRIPTION
**Pull request by bismurphy**
This is a really weird function. It has several calls to PlaySfx which use 3 arguments, for example `PlaySfx(0x64C, 0x30, 0);`.

There is one last Goto which I was not able to remove. Notice that this function has many, many return statements. My handling of these might be related to the remaining Goto. I generally, when possible, used structures like:
```
if(cond1){
    doThingA;
    return 1;
}
doThingB;
```
which could equally well be written as
```
if(cond1){
    doThingA;
    return 1;
} else {
    doThingB;
}
```
And it is possible that some rearranging of those if-else blocks would remove the Goto. It's a bit odd that the compiled code does a jump where it does.

I believe this function is responsible for much of Alucard's vertical acceleration management, especially what comes from player input. We do a bunch of checks of PAD_CROSS, PAD_UP, and PAD_DOWN. This function also does the check for PAD_TRIANGLE, and if it's pressed, calls func_8010FD88 which initiates the back-dash movement. 

I am open to name recommendations; something like VertButtonHandler might be good, though this function seems like a bit of a mish-mash of different factors with the player's movement.

### CODE REVIEW

File: src/dra/692E8.c [hunk 1]
```diff
@@ -869,8 +869,154 @@ void func_8010FD88(void) {
...
+    if (branchFlags & 8 && g_Player.unk46 == 0) {
+        func_8010E27C();
+    }
+    YAccel = -((branchFlags & 0x8000) != 0) & 0x2C00;
+    if (branchFlags & 0x10000) {
+        YAccel = 0x2C00;
+        if ((PLAYER.accelerationY + 0x1FFF) < 0x7FFFU &&
+            !(g_Player.unk44 & 0x20) && g_Player.padPressed & PAD_CROSS) {
+            YAccel = 0x8CC;
+        }
```
Comment (sonicdcer) [inline src/dra/692E8.c:887]:
```suggestion
if (branchFlags & 0x10000) {
        if ((PLAYER.accelerationY > -0x2000) && (PLAYER.accelerationY < 0x6000) && !(g_Player.unk44 & 0x20) && (g_Player.padPressed & PAD_CROSS)) {
            YAccel = 0x8CC;
        } else {
            YAccel = 0x2C00;
        }
    }

File: src/dra/692E8.c [hunk 2]
```diff
@@ -869,8 +869,154 @@ void func_8010FD88(void) {
...
+    }
+
+    if (branchFlags & 0x200) {
+        YAccel = 0x2C00;
+        if ((PLAYER.accelerationY + 0x1FFF) < 0x7FFFU) {
+            YAccel = 0x1600;
+        }
+    }
```
Comment (sonicdcer) [inline src/dra/692E8.c:895]:
```suggestion
if (branchFlags & 0x200) {
        if ((PLAYER.accelerationY > -0x2000) && (PLAYER.accelerationY < 0x6000)) {
            YAccel= 0x1600;
        } else {
            YAccel= 0x2C00;
        }
    }

File: src/dra/692E8.c [hunk 3]
```diff
@@ -869,8 +869,154 @@ void func_8010FD88(void) {
...
+
+    if (D_80097448 >= 0x29) {
+        tempYAccel = YAccel;
+        if (YAccel < 0) {
+            tempYAccel = YAccel + 3;
+        }
+        YAccel = tempYAccel >> 2;
+    }
```
Comment (sonicdcer) [inline src/dra/692E8.c:903]:
```suggestion
if (D_80097448 >= 41) {
        if (YAccel < 0) {
            tempYAccel = YAccel + 3;
        } else {
            tempYAccel = YAccel;
        }
        YAccel = tempYAccel >> 2;
    }

File: src/dra/692E8.c [hunk 4]
```diff
@@ -869,8 +869,154 @@ void func_8010FD88(void) {
...
+        PLAYER.accelerationY = -0x10000;
+    }
+
+    if (PLAYER.accelerationY >= 0) {
+        if (branchFlags & 1 && g_Player.pl_vram_flag & 1) {
+            if (g_Player.unk46 != 0) {
+                if ((g_Player.unk46 & 0x7FFF) == 0xFF) {
+                    func_8010E570(0);
+                    func_8010FAF4();
+                stupidplace:
```
Comment (sonicdcer) [inline src/dra/692E8.c:921]:
```suggestion
label:

File: src/dra/692E8.c [hunk 5]
```diff
@@ -869,8 +869,154 @@ void func_8010FD88(void) {
...
+                    PlaySfx(0x64C, 0x30, 0);
+                    return 1;
+                }
+
+                if (PLAYER.accelerationY > 0x6E000) {
+                    func_8010E470(1, 0);
+                    func_80134714(0x647);
+                    func_8011AAFC(g_CurrentEntity, 0U, 0);
```
Comment (sonicdcer) [inline src/dra/692E8.c:929]:
```suggestion
func_8011AAFC(g_CurrentEntity, 0, 0);

File: src/dra/692E8.c [hunk 6]
```diff
@@ -869,8 +869,154 @@ void func_8010FD88(void) {
...
+                    }
+                    PlaySfx(0x64C, 0x30, 0);
+                }
+
+                func_8010FAF4();
+                return 1;
+            }
+
+            if (PLAYER.accelerationY > 0x6E000) {
+                if ((PLAYER.step_s == 0x70) || (PLAYER.step == 4)) {
```
Comment (sonicdcer) [inline src/dra/692E8.c:944]:
```suggestion
if ((PLAYER.step_s == 112) || (PLAYER.step == 4)) {

File: src/dra/692E8.c [hunk 7]
```diff
@@ -869,8 +869,154 @@ void func_8010FD88(void) {
...
+                    func_8010E470(3, PLAYER.accelerationX / 2);
+                } else {
+                    func_8010E470(1, 0);
+                }
+                PlaySfx(0x647);
+                func_8011AAFC(g_CurrentEntity, 0U, 0);
```
Comment (sonicdcer) [inline src/dra/692E8.c:950]:
```suggestion
func_8011AAFC(g_CurrentEntity, 0, 0);

File: src/dra/692E8.c [hunk 8]
```diff
@@ -869,8 +869,154 @@ void func_8010FD88(void) {
...
+            }
+
+            if (g_Player.unk44 & 0x10) {
+                func_8010E6AC(1);
+                goto stupidplace;
+            }
+
+            if (ABS(PLAYER.accelerationX) > 0x20000) {
+                PlaySfx(0x647);
+                func_8011AAFC(g_CurrentEntity, 0U, 0);
```
Comment (sonicdcer) [inline src/dra/692E8.c:961]:
```suggestion
func_8011AAFC(g_CurrentEntity, 0, 0);

File: src/dra/692E8.c [hunk 9]
```diff
@@ -869,8 +869,154 @@ void func_8010FD88(void) {
     PlaySfx(NA_SE_AL_BACKSLIDE);
 }
 
-INCLUDE_ASM("asm/us/dra/nonmatchings/692E8", func_8010FDF8);
-s32 func_8010FDF8(/*?*/ s32);
+bool func_8010FDF8(s32 branchFlags) {
+    u8 stackpad[0x28];
```
Comment (sonicdcer) [inline src/dra/692E8.c:873]:
```suggestion
char pad[0x28];

**Review by sonicdcer — CHANGES_REQUESTED** [review]

**Comment by sonicdcer** [comment]
The PlaySfx arguments thing is very weird tho, I don't think there's another way to match this. Maybe the debug version of the function can take 3 args? I'd like to see the "HD" version of it

**Comment by Xeeynamo** [comment]
> The PlaySfx arguments thing is very weird tho, I don't think there's another way to match this. Maybe the debug version of the function can take 3 args? I'd like to see the "HD" version of it

I had a look. The HD version implementation is identical to the US counterpart.

File: src/dra/692E8.c (hunk 9 above)
Comment (bismurphy) [reply src/dra/692E8.c:873]: Why does it matter if it's u8 or char, and why does it matter if it's stackpad or pad?

Comment (sonicdcer) [reply src/dra/692E8.c:873]: See
https://discord.com/channels/1079389589950705684/1079395324960981062/1127503998321172481

**Review by Xeeynamo — APPROVED** [review]
I could not remove the label and I could not remove the extra parameters in `PlaySfx`. Everything else looks good. I will wait for further feedback from at least one of the other contributors before merging it, but to me is good enough to get it in :+1:

**Review by sozud — APPROVED** [review]

**Pull request merged**